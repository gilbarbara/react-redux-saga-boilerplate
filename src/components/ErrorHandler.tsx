import { Button, Container, Icon, Paragraph } from '@gilbarbara/components';

interface Props {
  error: Error;
  resetErrorBoundary: () => void;
}

export default function ErrorHandler({ error, resetErrorBoundary }: Props) {
  const handleClickReset = () => {
    resetErrorBoundary();
  };

  return (
    <Container data-component-name="ErrorHandler" fullScreen>
      <Icon name="danger-o" size={64} />
      <Paragraph my="md">{error.message}</Paragraph>

      <Button bg="red" onClick={handleClickReset}>
        Tentar novamente
      </Button>
    </Container>
  );
}
