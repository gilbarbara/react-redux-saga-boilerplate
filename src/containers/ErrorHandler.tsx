import React from 'react';
import { Container, Heading, Image } from 'styled-minimal';

type Props = {
  children?: any;
  onError?: (error: Error, componentStack: string) => void;
};

type ErrorInfo = {
  componentStack: string;
};

type State = {
  error: Error | null;
};

export default class ErrorHandler extends React.Component<Props, State> {
  public state: State = {
    error: null,
  };

  componentDidCatch(error: Error, info: ErrorInfo) {
    const { onError } = this.props;

    /* istanbul ignore else */
    if (typeof onError === 'function') {
      try {
        onError.call(this, error, info?.componentStack);
      } catch {
        // ignore
      }
    }

    this.setState({ error });
  }

  render() {
    const { children } = this.props;
    const { error } = this.state;

    if (error === null) {
      return children;
    }

    const message = error.toString();

    return (
      <Container bg="#000" cl="#fff" fullScreen textAlign="center" data-testid="ErrorHandler">
        <Image src={`${process.env.PUBLIC_URL}/media/images/error.gif`} alt="Error" width={128} />
        <Heading as="h3" my={3}>
          {message}
        </Heading>
      </Container>
    );
  }
}
