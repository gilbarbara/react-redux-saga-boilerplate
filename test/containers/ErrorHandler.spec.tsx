import React from 'react';

import ErrorHandler from 'containers/ErrorHandler';

import { render, screen } from 'test-utils';

const mockOnError = jest.fn();

describe('ErrorHandler', () => {
  const Throws = () => {
    throw new Error('Oh no!');
  };

  it('should handle the error', () => {
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => undefined);

    render(
      <ErrorHandler onError={mockOnError}>
        <Throws />
      </ErrorHandler>,
    );

    expect(screen.getByTestId('ErrorHandler')).toMatchSnapshot();
    expect(mockOnError).toHaveBeenCalledTimes(1);

    spy.mockRestore();
  });
});
