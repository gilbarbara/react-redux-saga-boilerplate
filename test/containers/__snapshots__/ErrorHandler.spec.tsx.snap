// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler should handle the error 1`] = `
.c0 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  padding-left: 16px;
  padding-right: 16px;
  position: relative;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 100vh;
  background-color: #000;
  color: #fff;
  text-align: center;
}

.c2 {
  box-sizing: border-box;
  font-size: 32px;
  font-family: inherit;
  font-weight: 700;
  line-height: 1;
  margin-bottom: 0;
  margin-top: 8px;
  margin-top: 16px;
  margin-bottom: 16px;
}

.c2:first-child {
  margin-top: 0;
}

.c1 {
  box-sizing: border-box;
  max-width: 100%;
  width: 128px;
}

@media screen and (min-width:1024px) {
  .c0 {
    padding-left: 32px;
    padding-right: 32px;
  }
}

<div
  class="c0"
  data-testid="ErrorHandler"
>
  <img
    alt="Error"
    class="c1"
    src="/media/images/error.gif"
    width="128"
  />
  <h3
    class="c2"
  >
    Error: Oh no!
  </h3>
</div>
`;
